### DISCUSSION TITLE
concatenating TransformParametersFiles

### DISCUSSION THREAD

**Discussion by amwink** [SuperElastix/elastix#840] [Q&A]
My script cycles over a number of subjects, all of whom have a PET image and an MR scan, each in their own space. Every PET needs to be registered to the MR, the MR to an MNI template, and then the PET to MNI via the MR.

So given 
```
PETFILE=<native-space PET>
PET_INPUT_DIR=<directory PETFILE>
PET_OUTPUT_DIR=<directory mni-space PET>
PET_OUTPUTFILE=<mni-space PET>
```
and
```
MRIFILE=<native-space T1w>
MRI_INPUT_DIR=<directory MRIFILE>
MRI_OUTPUT_DIR=<directory mni-space T1w>
MRI_OUTPUTFILE=<mni-space T1w>
TEMPLATE=$FSLDIR/data/standard/MNI152_T1_2mm.nii.gz
```
I first do
```
elastix -f $MRIFILE -m $PETFILE -out $PET_OUTPUT_DIR -p Parameters_Rigid.txt -p Parameters_Affine.txt
transformix -in $PETFILE -out $PET_OUTPUT_DIR -tp ${PET_OUTPUT_DIR}/TransformParameters.1.txt
mv $PET_OUTPUT_DIR/result.nii.gz ${PET_OUTPUTFILE//_pet.nii/_space-T1w_pet.nii}
```
and
```
elastix -f $TEMPLATE -m $MRIFILE -out $MRI_OUTPUT_DIR -p Parameters_Rigid.txt -p Parameters_Affine.txt -p Parameters_BSpline.txt
transformix -in $MRIFILE -out $MRI_OUTPUT_DIR -tp ${MRI_OUTPUT_DIR}/TransformParameters.2.txt
mv $MRI_OUTPUT_DIR/result.nii.gz ${MRI_OUTPUTFILE//_T1w.nii/_space-mni.nii} 
```
to register PET to MR and MR to MNI (the transformix calls are not strictly necessary but will be used later), both of which work as expected.

My cunning plan to warp the PET to MNI in 1 go was to change the line `(InitialTransformParametersFileName "NoInitialTransform")` in the MR rigid transform parameter file to  `(InitialTransformParametersFileName "${PET_OUTPUT_DIR}/TransformParameters.1.txt")` as explained in section 4.3 of the manual, so the initial parameters of the transform to warp the MR to MNI are now the final parameters for registering the PET to the MR. 
However, when I do that and apply the resulting transform to the PET:
```
sed $MRI_OUTPUT_DIR/TransformParameters.0.txt -e s#\"NoInitialTransform\"#\"${PET_OUTPUT_DIR}/TransformParameters.1.txt\"# > $MRI_OUTPUT_DIR/TransformParameters.0_PET.txt
transformix -in $PETFILE -out $PET_OUTPUT_DIR -tp ${MRI_OUTPUT_DIR}/TransformParameters.2.txt
mv $PET_OUTPUT_DIR/result.nii.gz ${PET_OUTPUTFILE//_pet.nii/_space-mni_pet.nii}			
```
the resulting is not registered correctly to the MNI template.

However, if instead I keep the MR transform parameters intact and apply them to the PET that was resampled in the MR space,
```
transformix -in ${PET_OUTPUTFILE//_pet.nii/_space-T1w_pet.nii} -out $PET_OUTPUT_DIR -tp ${MRI_OUTPUT_DIR}/TransformParameters.2.txt;
mv $PET_OUTPUT_DIR/result.nii.gz ${PET_OUTPUTFILE//_pet.nii/_desc-resampled2x_space-mni_pet.nii} 
```
the result is correctly registered to the MNI space.

Am I forgetting something for combining the two sets of `TransformParameters.txt` files?

Or could the problem be that using `TransformParameters.1.txt` from the PET-to-MR registration (affine) as an initial transform in `TransformParameters.0.txt` from the MR-to-MNI registration (rigid) discards a number of parameters?

**Comment by ntatsisk** [comment]
Hi @amwink, I converted your issue back to discussion.

Maybe you could find the following older discussions helpful?

- https://groups.google.com/g/elastix-imageregistration/c/ZpXWa6OJdT8/m/4aPs2CQwCQAJ
- https://groups.google.com/g/elastix-imageregistration/c/R0z5erk4qAQ/m/GYYwwSb7AgAJ

Let us know how it goes.

**Comment by amwink** [comment]
Ah thanks (I was trying to copy then delete the discussion but that did not work)!

I'm not sure that I understand.  

So if I have 2 files for a coregistration, let's say `CR_TransformParameters.0.txt` and `CR_TransformParameters.1.txt`. When I call `transformix`, I use `CR_TransformParameters.1.txt` (the affine transform) as the argument, which in turn specifies `CR_TransformParameters.0.txt` (the rigid transform) as its initial parameters, which itself specifies "NoInitialTransform".

Likewise for the standard space mapping, the argument to `transformix`, say `SSM_TransformParameters.2.txt` (the B-spline mapping) has `SSM_TransformParameters.1.txt` (affine) as its initial parameters, which in turn has `SSM_TransformParameters.0.txt`  (rigid) as initial parameters. That makes perfect sense.

So if I want to do a coregistration followed by a standard space mapping, using the same logic, then the standard space mapping should have the coregistration as its initial parameters, right? so in `SSM_TransformParameters.0.txt`, "NoInitialTransform" should be changed to `CR_TransformParameters.1.txt`.

The google message seems to suggest that it should be the other way round???

This is the way that I have connected the transforms, based on the manual and how the transforms inside the coregistration and the standard space mapping themselves were connected.

**Comment by amwink** [comment]
I have tried both ways now, both using `SSM_TransformParameters.2.txt` as initial parameters for `CR_TransformParameters.0.txt` and then calling transformix with `-tp CR_TransformParameters.1.txt`, and using `CR_TransformParameters.1.txt` as initial parameters for `SSM_TransformParameters.0.txt` and then calling transformix with `-tp SSM_TransformParameters.2.txt`. Neither of those give me theright result, that is, an image that looks like the result I get when I resample after coregistration, then resample again after standard-space mapping. Ideally, the extra interpolation, and the extra file, should not be necessary?

If anyone has a worked example I would be interested to see that, thanks!
